README.md

```diff
@@ -1,198 +1,48 @@
-		* [P0] Make source{d} the face of #MLonCode ([#50](https://github.com/src-d/okrs/issues/50))
-			* [P0] Develop half a day workshop on getting started with #MLonCode with source{d} stack (PGA, gitbase?, engine, etc)
-			* [P1] Run that workshop at least once at a meetup / conference
-			* [P0] Create a new talk on the future of #MLonCode and deliver it at least four times (existing confs and meetups)
-			* [P1] Make that workshop and talk available online
-			* [P1] Create two demos showing the power of our stack solving a “real problem”
-			* [P2] Determine what percentage of time source{d} engineers should spend working on docs/tutorials/talks/etc
-		* [P0] Make source{d} the lead of MLonCode research ([#42](https://github.com/src-d/okrs/issues/42))
-			* [P0] Create and maintain a research log with TL;DR summaries of research read
-			* [P1] Publish 6 ML blog posts in Q2
-			* [P2] Speak about ML on >4 relevant conferences
-			* [P2] Reach 1,000 cumulative upvotes of new blog posts on Reddit
-		* [P0] Enable sharing and publishing MLonCode models ([#44](https://github.com/src-d/okrs/issues/44))
-			* [P0] Define and design the product for model sharing (relates to
-			* [P1] Conduct user research with all key users (e.g.: key academic groups, users of our stack).
-		* [P0] Release updated content for sourced.tech ([#13](https://github.com/src-d/okrs/issues/13))
-			* [P0] Create a content strategy plan ([#35](https://github.com/src-d/okrs/issues/35))
-				* [P0] Editorial Environment Setup
-				* [P0] Customer Journeys
-				* [P0] Content Model
-				* [P0] Copywriting
-				* [P0] Metrics & Evaluation Analytics Setup
-				* [P1] Information Architecture
-				* [P2] Establish a Content Management Plan
-			* [P1] Update brand touchpoints according to content plan
-			* [P1] Apply new Visual Style to sourced.tech ([#32](https://github.com/src-d/okrs/issues/32))
-				* [P0] Visual Artifacts Study
-				* [P1] Research
-				* [P1] Chromatic Palete Study
-				* [P1] Typography Study
-				* [P1] Initial Proposals
-				* [P1] Pure css update PR
-			* [P1] Publish the CAT case-study development post on Medium ([#38](https://github.com/src-d/okrs/issues/38))
-				* [P0] Write a blog post regarding the process, methodologies and collaboration flow between teams on CAT
-	* Define DevRel organization ([#48](https://github.com/src-d/okrs/issues/48))
-		* [P0] Describe the metrics of devrel success
-		* [P0] Set monitoring up for all processes valuable to devrel
-		* [P1] Create a dashboard displaying those metrics
-		* [P0] Implement kanban processes in DevRel, follow the source{d} Engineering Methodology
-		* [P0] Set up biweekly sync-ups with Eiso, Marcelo, Maximo, and Fernanda
+# OKRs 2018 Q3
+
+## Bring 'Code as Data' to market with enterprises
+
+### gitbase & engine
+
+- [ ] Stable & Performance Release
+- [ ] Integration of gitbase with engine
+- [ ] New engine interface based on gitbase
+
+### Babelfish
+
+- [ ] C/C++ driver
```

Comment:
I don't think it's realistic for Q3.

Let's focus on increasing semantic object coverage for drivers (type info and classes are good next milestone) as well as better documentation and user experience in general.

We still have Java/JS which are very popular, so focus on making them work great should have a higher priority, I think. As an option, we can look at bringing C# support instead that might cover some enterprise use cases.

LA will definitely start looking into C/C++, but stating that it will be ready for the end of Q3 is too optimistic even for OKRs.

Reply:
Actually, we'll have to prioritize C# (much easier), C & C++ since without it we can't deliver an enterprise product. I would argue that it's the highest priority item for LA, but am fully aware that we are likely not to deliver C/C++ in Q3.

Reply:
C++ is a driver that we could definitely have in Q3 (I have a half written but severely outdated native driver) **IF** we take some pragmatic shortcuts. The most important one is to not try to parse the branching preprocessor directives and macros but instead get the values from the environment (or the defaults) or allow the user to provide values if needed and work over the result of the preprocessed code. And for running this (pragmatic shortcut number 2) this "preprocessor runner" would need access to the full repository, and maybe a docker image with a build environment and the the 500 most popular C++ libraries installed (this still would not warrant 100% of code working, but I think a high % is achievable). 

This (running the preprocessor before parsing the resulting code) is what a guy from Facebook that was in this year's FOSDEM dinner (sorry, forgot the name) told me they do themselves. Once this is done, CDT's parser works pretty well even for incomplete information.

Reply:
@mcuadros @smola what do you think of this approach? 

From a product point of view, we can't be without a C/C++ driver for much longer, so I am open to this short cut.

Reply:
If I understood correctly, access to the full repository is not something that can happen only at the driver side, it would need to review most of babelfish design assumptions.

For the library stuff, I assume we (or customers) can use custom builds of the C/C++ driver for a given environment, as you suggest. I think that would be an acceptable tradeoff.

Reply:
The approach to follow, it the one used by the rest of the community, the macros are entities of the language, they are not expended or interpreted in any way. Making in this way I think is archivable in Q3

Reply:
The really problematic things are branching directives (#ifdef, #ifndef, etc). #define for constants and macros are not really problematic, trough parsing the macro bodies could be hairy in some cases, but it could be avoided and just tag the usage of them as "function call" which is actually what CDT does by default.